ARCs/arc-0062.md

```diff
@@ -0,0 +1,333 @@
+
+Creating an ASA on the ledger _does not_ imply its `total` supply is immediately
+“minted” or “circulating”. In fact, the semantic of token “minting” on Algorand is
+slightly different from other blockchains: it is not coincident with the token units
+creation on the ledger.
+
+The Reserve Address, one of the 4 addresses of ASA Role-Based-Access-Control (RBAC),
+is conventionally used to identify the portion of `total` supply not yet in circulation.
+The Reserve Address has no “privilege” over the token: it is just a “logical” label
+used (client-side) to classify an existing amount of ASA as “not in circulation”.
+
+According to this convention, “minting” an amount of ASA units is equivalent to
+_moving that amount out of the Reserve Address_.
+
+> ASA may have the Reserve Address assigned to a Smart Contract to enforce specific
+> “minting” policies, if needed.
+
+This convention led to a simple and unsophisticated semantic of ASA circulating
+supply, widely adopted by clients (wallets, explorers, etc.) to provide standard
+information:
+
+```text
+circulating_supply = total - reserve_balance
+```
+
+Where `reserve_balance` is the ASA balance hold by the Reserve Address.
+
+However, the simplicity of such convention, who fostered adoption across the Algorand
+ecosystem, poses some limitations. Complex and sophisticated use-cases of ASA, such
+as regulated stable-coins and tokenized securities among others, require more
+detailed and expressive definitions of circulating supply.
+
+As an example, an ASA could have “burned”, “locked” or “pre-minted” amounts of token,
+not held in the Reserve Address, which _should not_ be considered as “circulating”
+supply. This is not possible with the basic ASA protocol convention.
+
+This ARC proposes a standard ABI _read-only_ method (getter) to provide the circulating
+supply of an ASA.
+
+## Specification
+
+The keywords "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**",
+"**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**"
+in this document are to be interpreted as described in <a href="https://datatracker.ietf.org/doc/html/rfc2119">RFC 2119</a>.
+
+> Notes like this are non-normative.
+
+### ABI Method
+
+A compliant ASA, whose circulating supply definition conforms to this ARC, **MUST**
+implement the following method on an Application (referred as _Circulating Supply
+App_ in this specification):
+
+```json
+{
+  "name": "arc62_get_circulating_supply",
+  "readonly": true,
+  "args": [
+    {
+      "type": "uint64",
+      "name": "asset_id",
+      "desc": "ASA ID of the circulating supply"
+    }
+  ],
+  "returns": {
+    "type": "uint64",
+    "desc": "ASA circulating supply"
+  },
+  "desc": "Get ASA circulating supply"
+}
+```
+
+The `arc62_get_circulating_supply` **MUST** be a _read-only_ ([ARC-22](./arc-0022.md))
+method (getter).
+
+### Usage
+
+Getter calls **SHOULD** be _simulated_.
```

Comment:
To enable the use case in which getter calls are made on-chain by another application (and not just simulated), would it be reasonable to include the following statement?

> The `arcXXX_get_circulating_supply` method **SHOULD** not require more foreign references than can be provided on a single applicationCall transaction within the constraints of the Algorand Virtual Machine.

This also raises another question, which is: Should the ARC require that its implementations specify the minimum AVM version required to call the method?

Reply:
I think this is an implementation detail. I would rather add a recommendation for the reference implementation to use an AVM version that supports resource pooling. The specification of the method interface does not really care about _how_ resources are populated (e.g. they could be added to the AppCall or provided somehow within the same Group Transaction).